File: packages/nitro-renderer/src/room/object/visualization/furniture/FurnitureVisualizationData.ts
import { type IAssetColor, type IAssetData, type IAssetVisualizationData, type IAssetVisualizationDirection, type IAssetVisualizationLayer, type IObjectVisualizationData, RoomGeometryScaleType } from '@nitrodevco/nitro-api';
import type { BLEND_MODES } from 'pixi.js';

import { ColorData, LayerData, SizeData } from '../data';

export class FurnitureVisualizationData implements IObjectVisualizationData {
    public static LAYER_LETTERS: string[] = [
        'a',
        'b',
        'c',
        'd',
        'e',
        'f',
        'g',
        'h',
        'i',
        'j',
        'k',
        'l',
        'm',
        'n',
        'o',
        'p',
        'q',
        'r',
        's',
        't',
        'u',
        'v',
        'w',
        'x',
        'y',
        'z',
    ];

    private _type: string = '';
    private _sizes: number[] = [];
    private _sizeDatas: Map<number, SizeData> = new Map();
    private _lastSize: number = -1;
    private _lastSizeScale: RoomGeometryScaleType = -1;
    private _lastSizeData: SizeData | undefined = undefined;
    private _lastSizeDataScale: RoomGeometryScaleType = -1;

    public initialize(asset: IAssetData | undefined): boolean {
        this.reset();

        if (!asset) return false;

        this._type = asset.type;

        if (!asset.visualizations || !this.defineVisualizations(asset.visualizations)) {
            this.reset();

            return false;
        }

        return true;
    }

    public dispose(): void {
        if (this._sizeDatas && this._sizeDatas.size) {
            for (const size of this._sizeDatas.values()) size?.dispose();
        }

        this._lastSizeData = undefined;
        this._sizes = [];
    }

    private reset(): void {
        this._type = '';

        if (this._sizeDatas && this._sizeDatas.size) {
            for (const size of this._sizeDatas.values()) size?.dispose();
        }

        this._sizeDatas.clear();

        this._sizes = [];
        this._lastSizeData = undefined;
        this._lastSizeDataScale = -1;
    }

    protected createSizeData(scale: RoomGeometryScaleType, layerCount: number, angle: number): SizeData {
        return new SizeData(layerCount, angle);
    }

    protected defineVisualizations(visualizations: IAssetVisualizationData[]): boolean {
        if (!visualizations) return false;

        for (const visualization of visualizations) {
            const layerCount = visualization.layerCount ?? 0;
            const angle = visualization.angle ?? 45;

            let size = visualization.size ?? 1;

            if (size < 1) size = 1;

            if (this._sizeDatas.get(size)) return false;

            const sizeData = this.createSizeData(size, layerCount, angle);

            if (!sizeData) return false;

            for (const key in visualization) {
                if (!this.processVisualElement(sizeData, key, visualization[key] as object)) {
                    sizeData.dispose();

                    return false;
                }
            }

            this._sizeDatas.set(size, sizeData);

            this._sizes.push(size);
        }

        this.removeInvalidSizes();

        this._sizes.sort();

        return true;
    }

    private removeInvalidSizes(): void {
        if (!this._sizes || !this._sizes.length) return;

        const zoomedIn = this._sizeDatas.get(RoomGeometryScaleType.ZoomedIn);
        const zoomedOut = this._sizeDatas.get(RoomGeometryScaleType.ZoomedOut);

        if (zoomedIn && zoomedOut) {
            if (zoomedIn.layerCount !== zoomedOut.layerCount) {
                this._sizeDatas.delete(RoomGeometryScaleType.ZoomedOut);

                const index = this._sizes.indexOf(RoomGeometryScaleType.ZoomedOut);

                if (index >= 0) this._sizes.splice(index, 1);
            }
        }
    }

    protected processVisualElement(sizeData: SizeData, key: string, data: object): boolean {
        if (!sizeData || !key || !data) return false;

        switch (key) {
            case 'layers':
                if (!sizeData.processLayers(data as IAssetVisualizationLayer[])) return false;
                break;
            case 'directions':
                if (!sizeData.processDirections(data as IAssetVisualizationDirection[])) return false;
                break;
            case 'colors':
                if (!sizeData.processColors(data as IAssetColor[])) return false;
                break;
        }

        return true;
    }

    public getValidSize(scale: RoomGeometryScaleType): number {
        if (scale === this._lastSizeScale) return this._lastSize;

        const sizeIndex = this.getSizeIndex(scale);

        let newScale = -1;

        if (sizeIndex < this._sizes.length) newScale = this._sizes[sizeIndex];

        this._lastSizeScale = scale;
        this._lastSize = newScale;

        return newScale;
    }

    private getSizeIndex(size: number): number {
        if (size <= 0) return 0;

        let index = 0;
        let iterator = 1;

        while (iterator < this._sizes.length) {
            if (this._sizes[iterator] > size) {
                if (this._sizes[iterator] / size < size / this._sizes[iterator - 1]) index = iterator;

                break;
            }

            index = iterator;

            iterator++;
        }

        return index;
    }

    protected getSizeData(size: RoomGeometryScaleType): SizeData | undefined {
        if (size === this._lastSizeDataScale) return this._lastSizeData;

        const sizeIndex = this.getSizeIndex(size);

        if (sizeIndex < this._sizes.length) this._lastSizeData = this._sizeDatas.get(this._sizes[sizeIndex]);
        else this._lastSizeData = undefined;

        this._lastSizeDataScale = size;

        return this._lastSizeData;
    }

    public getLayerCount(scale: RoomGeometryScaleType): number {
        return this.getSizeData(scale)?.layerCount ?? LayerData.DEFAULT_COUNT;
    }

    public getValidDirection(scale: RoomGeometryScaleType, direction: number): number {
        return this.getSizeData(scale)?.getValidDirection(direction) ?? LayerData.DEFAULT_DIRECTION;
    }

    public getLayerTag(scale: RoomGeometryScaleType, direction: number, layerId: number): string {
        return this.getSizeData(scale)?.getLayerTag(direction, layerId) ?? LayerData.DEFAULT_TAG;
    }

    public getLayerBlendMode(scale: RoomGeometryScaleType, direction: number, layerId: number): BLEND_MODES {
        return this.getSizeData(scale)?.getLayerBlendMode(direction, layerId) ?? LayerData.DEFAULT_BLEND_MODE;
    }

    public getLayerAlpha(scale: RoomGeometryScaleType, direction: number, layerId: number): number {
        return this.getSizeData(scale)?.getLayerAlpha(direction, layerId) ?? LayerData.DEFAULT_ALPHA;
    }

    public getLayerColor(scale: RoomGeometryScaleType, layerId: number, colorId: number): number {
        return this.getSizeData(scale)?.getLayerColor(layerId, colorId) ?? ColorData.DEFAULT_COLOR;
    }

    public getLayerIgnoreMouse(scale: RoomGeometryScaleType, direction: number, layerId: number): boolean {
        return this.getSizeData(scale)?.getLayerIgnoreMouse(direction, layerId) ?? LayerData.DEFAULT_IGNORE_MOUSE;
    }

    public getLayerXOffset(scale: RoomGeometryScaleType, direction: number, layerId: number): number {
        return this.getSizeData(scale)?.getLayerXOffset(direction, layerId) ?? LayerData.DEFAULT_XOFFSET;
    }

    public getLayerYOffset(scale: RoomGeometryScaleType, direction: number, layerId: number): number {
        return this.getSizeData(scale)?.getLayerYOffset(direction, layerId) ?? LayerData.DEFAULT_YOFFSET;
    }

    public getLayerZOffset(scale: RoomGeometryScaleType, direction: number, layerId: number): number {
        return this.getSizeData(scale)?.getLayerZOffset(direction, layerId) ?? LayerData.DEFAULT_ZOFFSET;
    }

    public get type(): string {
        return this._type;
    }
}

File: packages/nitro-swift/Sources/NitroRoom/Furniture/FurnitureVisualizationData.swift
import Foundation

/// Swift port of `FurnitureVisualizationData` (packages/nitro-renderer/src/room/object/visualization/furniture/FurnitureVisualizationData.ts),
/// merged with what the TS source splits into a separate `FurnitureAnimatedVisualizationData`
/// subclass (whose only differences are: always constructing `AnimationSizeData` instead of plain
/// `SizeData`, routing the `animations` key to `AnimationSizeData.defineAnimations`, and a handful
/// of animation-table pass-through accessors). Since Swift has no per-furniture-type visualization
/// factory in this port (see `FurnitureAnimatedVisualization`'s doc comment), the choice is made
/// here instead, at parse time: if *any* declared size has an `animations` array, every size for
/// this furniture gets an `AnimationSizeData` (matching `FurnitureAnimatedVisualizationData.createSizeData`
/// always doing so, since in the original the visualization-data *class* - not the per-size data -
/// determines this), so the animation pass-through methods below stay valid across every size.
public final class FurnitureVisualizationData {
    public static let layerLetters: [String] = (0..<26).map { String(UnicodeScalar(UInt8(97 + $0))) }

    public private(set) var type: String = ""
    /// True if any declared size for this furniture had an `animations` array - see the class doc
    /// comment for why this (not the visualization-data *class*, as in the TS original) is what
    /// decides whether `AnimationSizeData` is used. `FurnitureVisualizationFactory` uses this to
    /// choose between a plain `FurnitureVisualization` and a `FurnitureAnimatedVisualization`.
    public private(set) var isAnimated: Bool = false

    private var sizes: [Int] = []
    private var sizeDatas: [Int: SizeData] = [:]

    private var lastSize: Int = -1
    private var lastSizeScale: Int = -1
    private var lastSizeData: SizeData?
    private var lastSizeDataScale: Int = -1

    public init() {}

    @discardableResult
    public func initialize(type: String, visualizations: [AssetVisualizationData]?) -> Bool {
        reset()

        self.type = type

        guard let visualizations, defineVisualizations(visualizations) else {
            reset()

            return false
        }

        return true
    }

    private func reset() {
        type = ""
        isAnimated = false
        sizeDatas.removeAll()
        sizes.removeAll()
        lastSizeData = nil
        lastSizeDataScale = -1
    }

    private func defineVisualizations(_ visualizations: [AssetVisualizationData]) -> Bool {
        let animated = visualizations.contains { $0.animations != nil }

        isAnimated = animated

        for visualization in visualizations {
            let layerCount = visualization.layerCount ?? 0
            let angle = visualization.angle ?? 45
            let size = max(visualization.size ?? 1, 1)

            if sizeDatas[size] != nil { return false }

            let sizeData: SizeData = animated ? AnimationSizeData(layerCount: layerCount, angle: angle) : SizeData(layerCount: layerCount, angle: angle)

            if let layers = visualization.layers { sizeData.processLayers(layers) }
            if let directions = visualization.directions { sizeData.processDirections(directions) }
            if let colors = visualization.colors { sizeData.processColors(colors) }

            if let animationSizeData = sizeData as? AnimationSizeData, let animations = visualization.animations {
                if !animationSizeData.defineAnimations(animations) { return false }
            }

            sizeDatas[size] = sizeData
            sizes.append(size)
        }

        removeInvalidSizes()
        sizes.sort()

        return true
    }

    private func removeInvalidSizes() {
        guard !sizes.isEmpty else { return }

        guard
            let zoomedIn = sizeDatas[RoomGeometryScaleType.zoomedIn.rawValue],
            let zoomedOut = sizeDatas[RoomGeometryScaleType.zoomedOut.rawValue]
        else { return }

        if zoomedIn.layerCount != zoomedOut.layerCount {
            sizeDatas.removeValue(forKey: RoomGeometryScaleType.zoomedOut.rawValue)
            sizes.removeAll { $0 == RoomGeometryScaleType.zoomedOut.rawValue }
        }
    }

    public func getValidSize(_ scale: Int) -> Int {
        if scale == lastSizeScale { return lastSize }

        let index = getSizeIndex(scale)
        let newScale = index < sizes.count ? sizes[index] : -1

        lastSizeScale = scale
        lastSize = newScale

        return newScale
    }

    /// "Nearest by ratio, not by absolute distance" - meaningful because furniture sizes are
    /// scale factors (1/32/64), not a linear range.
    private func getSizeIndex(_ size: Int) -> Int {
        guard size > 0 else { return 0 }

        var index = 0
        var iterator = 1

        while iterator < sizes.count {
            if sizes[iterator] > size {
                if Double(sizes[iterator]) / Double(size) < Double(size) / Double(sizes[iterator - 1]) { index = iterator }

                break
            }

            index = iterator
            iterator += 1
        }

        return index
    }

    private func getSizeData(_ size: Int) -> SizeData? {
        if size == lastSizeDataScale { return lastSizeData }

        let index = getSizeIndex(size)

        lastSizeData = index < sizes.count ? sizeDatas[sizes[index]] : nil
        lastSizeDataScale = size

        return lastSizeData
    }

    public func getLayerCount(_ scale: Int) -> Int { getSizeData(scale)?.layerCount ?? LayerData.defaultCount }

    public func getValidDirection(_ scale: Int, _ direction: Int) -> Int {
        getSizeData(scale)?.getValidDirection(direction) ?? LayerData.defaultDirection
    }

    public func getLayerTag(_ scale: Int, _ direction: Int, _ layerId: Int) -> String {
        getSizeData(scale)?.getLayerTag(direction, layerId) ?? LayerData.defaultTag
    }

    public func getLayerBlendMode(_ scale: Int, _ direction: Int, _ layerId: Int) -> String {
        getSizeData(scale)?.getLayerBlendMode(direction, layerId) ?? LayerData.defaultBlendMode
    }

    public func getLayerAlpha(_ scale: Int, _ direction: Int, _ layerId: Int) -> Double {
        getSizeData(scale)?.getLayerAlpha(direction, layerId) ?? LayerData.defaultAlpha
    }

    public func getLayerColor(_ scale: Int, _ layerId: Int, _ colorId: Int) -> UInt32 {
        getSizeData(scale)?.getLayerColor(layerId, colorId) ?? ColorData.defaultColor
    }

    public func getLayerIgnoreMouse(_ scale: Int, _ direction: Int, _ layerId: Int) -> Bool {
        getSizeData(scale)?.getLayerIgnoreMouse(direction, layerId) ?? LayerData.defaultIgnoreMouse
    }

    public func getLayerXOffset(_ scale: Int, _ direction: Int, _ layerId: Int) -> Double {
        getSizeData(scale)?.getLayerXOffset(direction, layerId) ?? LayerData.defaultXOffset
    }

    public func getLayerYOffset(_ scale: Int, _ direction: Int, _ layerId: Int) -> Double {
        getSizeData(scale)?.getLayerYOffset(direction, layerId) ?? LayerData.defaultYOffset
    }

    public func getLayerZOffset(_ scale: Int, _ direction: Int, _ layerId: Int) -> Double {
        getSizeData(scale)?.getLayerZOffset(direction, layerId) ?? LayerData.defaultZOffset
    }

    // MARK: - Animation pass-through (ports of `FurnitureAnimatedVisualizationData`'s methods)

    private func getAnimationSizeData(_ scale: Int) -> AnimationSizeData? { getSizeData(scale) as? AnimationSizeData }

    public func hasAnimation(_ scale: Int, _ animationId: Int) -> Bool {
        getAnimationSizeData(scale)?.hasAnimation(animationId) ?? false
    }

    public func getAnimationCount(_ scale: Int) -> Int {
        getAnimationSizeData(scale)?.getAnimationCount() ?? 0
    }

    public func getAnimationId(_ scale: Int, _ animationId: Int) -> Int {
        getAnimationSizeData(scale)?.getAnimationId(animationId) ?? 0
    }

    public func isImmediateChange(_ scale: Int, _ animationId: Int, _ state: Int) -> Bool {
        getAnimationSizeData(scale)?.isImmediateChange(animationId, state) ?? false
    }

    public func getStartFrame(_ scale: Int, _ animationId: Int, _ direction: Int) -> Double {
        getAnimationSizeData(scale)?.getStartFrame(animationId, direction) ?? 0
    }

    public func getFrame(_ scale: Int, _ animationId: Int, _ direction: Int, _ layerId: Int, _ frameCount: Double) -> AnimationFrame? {
        getAnimationSizeData(scale)?.getFrame(animationId, direction, layerId, frameCount)
    }

    public func getFrameFromSequence(
        _ scale: Int, _ animationId: Int, _ direction: Int, _ layerId: Int, _ sequenceId: Int, _ offset: Int, _ frameCount: Double
    ) -> AnimationFrame? {
        getAnimationSizeData(scale)?.getFrameFromSequence(animationId, direction, layerId, sequenceId, offset, frameCount)
    }
}
